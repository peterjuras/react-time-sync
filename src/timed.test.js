import { mount } from "enzyme";

import React from "react";
import Timed from "./timed";
import TimeSync from "time-sync";
import lolex from "lolex";
import TimeProvider from "./time-provider";

describe("#Timed", () => {
  let clock;

  beforeEach(() => {
    clock = lolex.install({ now: 1 });
  });

  afterEach(() => {
    clock.uninstall();
  });

  it("should be exported correctly", () => expect(Timed).toBeDefined());

  it("should mount and unmount correctly", () => {
    const ref = mount(
      <TimeProvider>
        <Timed>{() => <div />}</Timed>
      </TimeProvider>
    );
    ref.unmount();
  });

  it("should throw if context is not found", () => {
    expect(() => {
      const ref = mount(<Timed>{() => <div />}</Timed>);
      ref.unmount();
    }).toThrowErrorMatchingSnapshot();
  });

  it("should respect prop updates", () => {
    let renderCalledCount = 0;

    const ref = mount(
      <TimeProvider>
        <Timed>
          {({ currentTime }) => {
            renderCalledCount++;
            return <div>{currentTime}</div>;
          }}
        </Timed>
      </TimeProvider>
    );

    expect(ref).toMatchSnapshot();

    clock.tick(999);
    ref.update();
    expect(ref).toMatchSnapshot();

    clock.tick(1000);
    ref.update();
    expect(ref).toMatchSnapshot();

    const newProps = {
      unit: 5,
      interval: TimeSync.MINUTES
    };

    ref.setProps({
      children: React.cloneElement(ref.props().children, newProps)
    });

    ref.update();
    expect(ref).toMatchSnapshot();

    clock.tick(1000 * 5 * 60);

    ref.update();
    expect(ref).toMatchSnapshot();

    ref.unmount();

    expect(renderCalledCount).toBe(5);
  });

  it("should not update when until number is reached", () => {
    clock.tick(999);

    let renderCalledCount = 0;

    const ref = mount(
      <TimeProvider>
        <Timed until={1}>
          {({ currentTime, finished }) => {
            renderCalledCount++;
            return (
              <div>
                {currentTime}
                {`${finished}`}
              </div>
            );
          }}
        </Timed>
      </TimeProvider>
    );

    expect(ref).toMatchSnapshot();

    clock.tick(1000);

    ref.update();
    expect(ref).toMatchSnapshot();

    ref.unmount();

    expect(renderCalledCount).toBe(1);
  });

  it("should not update when until function returns true", () => {
    clock.tick(999);

    let renderCalledCount = 0;

    const ref = mount(
      <TimeProvider>
        <Timed until={currentTime => currentTime === 5}>
          {({ currentTime, finished }) => {
            renderCalledCount++;
            return (
              <div>
                {currentTime}
                {`${finished}`}
              </div>
            );
          }}
        </Timed>
      </TimeProvider>
    );

    expect(ref).toMatchSnapshot();

    clock.tick(10000);

    ref.update();
    expect(ref).toMatchSnapshot();

    ref.unmount();

    expect(renderCalledCount).toBe(5);
  });
});
