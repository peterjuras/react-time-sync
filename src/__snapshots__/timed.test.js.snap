// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Timed should not update when until function returns true 1`] = `
<TimeProvider>
  <TimedWrapper
    until={[Function]}
  >
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval={null}
      unit={null}
      until={[Function]}
    >
      <div>
        1
        false
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should not update when until function returns true 2`] = `
<TimeProvider>
  <TimedWrapper
    until={[Function]}
  >
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval={null}
      unit={null}
      until={[Function]}
    >
      <div>
        5
        true
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should not update when until number is reached 1`] = `
<TimeProvider>
  <TimedWrapper
    until={1}
  >
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval={null}
      unit={null}
      until={1}
    >
      <div>
        1
        true
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should not update when until number is reached 2`] = `
<TimeProvider>
  <TimedWrapper
    until={1}
  >
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval={null}
      unit={null}
      until={1}
    >
      <div>
        1
        true
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should respect prop updates 1`] = `
<TimeProvider>
  <TimedWrapper>
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval={null}
      unit={null}
      until={null}
    >
      <div>
        0
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should respect prop updates 2`] = `
<TimeProvider>
  <TimedWrapper>
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval={null}
      unit={null}
      until={null}
    >
      <div>
        1
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should respect prop updates 3`] = `
<TimeProvider>
  <TimedWrapper>
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval={null}
      unit={null}
      until={null}
    >
      <div>
        2
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should respect prop updates 4`] = `
<TimeProvider>
  <TimedWrapper
    interval="m"
    unit={5}
  >
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval="m"
      unit={5}
      until={null}
    >
      <div>
        0
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should respect prop updates 5`] = `
<TimeProvider>
  <TimedWrapper
    interval="m"
    unit={5}
  >
    <Timed
      $$_TIME_SYNC_HIDDEN_$$={
        Object {
          "addTimer": [Function],
          "getCurrentTime": [Function],
        }
      }
      interval="m"
      unit={5}
      until={null}
    >
      <div>
        300
      </div>
    </Timed>
  </TimedWrapper>
</TimeProvider>
`;

exports[`#Timed should throw if context is not found 1`] = `"Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"`;
